bundle/src/main/java/com/adobe/acs/tools/explain_query/impl/ExplainQueryServlet.java

```diff
@@ -406,25 +451,86 @@ private static boolean checkMDCSupport(BundleContext context) {
         return true;
     }
 
-    private static class QueryLogCollector extends TurboFilter {
-        //MDC key defined in org.apache.jackrabbit.oak.query.QueryEngineImpl
-        private static final String QUERY_ANALYZE = "oak.query.analyze";
+    /**
+     * TurboFilters causes slowness as they are executed on critical path
+     * Hence care is taken to only register the filter only when required
+     * Logic below ensures that filter is only registered for the duration
+     * or request which needs to be "monitored".
+     *
+     * If multiple such request are performed then also only one filter gets
+     * registered
+     */
+    private void registerLogCollector() {
+        synchronized (logCollectorRegCount) {
+            int count = logCollectorRegCount.getAndIncrement();
+            if (count == 0) {
+                ServiceRegistration reg = bundleContext.registerService(TurboFilter.class.getName(),
+                        this.logCollector, null);
+                logCollectorReg.set(reg);
+            }
+        }
+    }
+
+    private void unregisterLogCollector() {
+        synchronized (logCollectorRegCount) {
+            int count = logCollectorRegCount.decrementAndGet();
+            if (count == 0) {
+                ServiceRegistration reg = logCollectorReg.getAndSet(null);
+                reg.unregister();
+            }
+        }
+    }
+
+    @Activate
+    private void activate(Map<String, ?> config, BundleContext context) {
+        this.bundleContext = context;
+
+        this.loggers = OsgiPropertyUtil.toMap(PropertiesUtil.toStringArray(
+                config.get(PROP_LOGGER_NAMES), new String[0]), "=", true, null);
+
+        if (loggers != null && !loggers.isEmpty()) {
+            this.pattern = PropertiesUtil.toString(config.get(PROP_MSG_PATTERN), DEFAULT_PATTERN);
+            this.msgCountLimit = PropertiesUtil.toInteger(config.get(PROP_LOG_COUNT_LIMIT), DEFAULT_LIMIT);
+
+            // Create a single logCollector; this will be used as TurboFilters are register with the execution of
+            // this Servlet
+            this.logCollector = new QueryLogCollector(loggers, pattern, msgCountLimit, checkMDCSupport(context));
```

Comment:
Creating QueryLogCollector once per @Activate; this same instance is used whenever the TurboFilter is register.